### PR TITLE
Disconnect corelog properly when core device connection failed

### LINKED ISSUE #1897: Make artiq handle coredevice power cycles gracefully
# ARTIQ Feature Request
## Problem this request addresses

At the moment, power cycling the core device while running an artiq master and all the trimmings seems to require restarting the master and any dashboards, for a couple of reasons:

* moninj proxy doesn't reconnect to core device
* haven't checked but I'm assuming core logging doesn't reconnect to core device either
* even if you restart the moninj proxy manually, the dashboard won't try to reconnect to the newly started proxy

This seems like it shouldn't be necessary, and involves some annoying overheads like waiting for the master to scan the experiment directory again.

## Describe the solution you'd like

A dashboard and/or artiq_client based trigger that tells the master to reconnect moninj proxy and core logging, and sends a broadcast signal to dashboards to reconnect to the moninj proxy.

## Additional context

I'm not sure how often other users are power cycling their core device; at the moment we are running the bleeding edge to get new features so we typically have to reset at least once per day on our live system, and many times per day on the test setup used for debugging, so this amounts to a reasonable amount of time for us.

### PR DESCRIPTION
**Pull request by Deepskyhunter**
# ARTIQ Pull Request

## Description of Changes
Replace asyncio.open_connection with async_open_connection that
has keepalive
Put get_logs_task inside asyncio.wait() for termination when
disconnected

## Type of Changes

|   | Type |
| ------------- | ------------- |
| ✓  | :bug: Bug fix  |

### All Pull Requests

- [x] Use correct spelling and grammar.
- [x] Close/update issues.

### Code Changes

- [x] Run `flake8` to check code style (follow PEP-8 style). `flake8` has issues with parsing Migen/gateware code, ignore as necessary.
- [x] Test your changes or have someone test them. Mention what was tested and how.

### Git Logistics

- [x] Split your contribution into logically separate changes (`git rebase --interactive`). Merge/squash/fixup commits that just fix or amend previous commits. Remove unintended changes & cleanup. See [tutorial](https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase).
- [x] Write short & meaningful commit messages. Review each commit for messages (`git show`). Format:
  ```
  topic: description. < 50 characters total.
  
  Longer description. < 70 characters per line
  ```

### Licensing

See [copyright & licensing for more info](https://github.com/m-labs/artiq/blob/master/CONTRIBUTING.rst#copyright-and-sign-off).
ARTIQ files that do not contain a license header are copyrighted by M-Labs Limited and are licensed under LGPLv3+.

### CODE REVIEW

**Comment by sbourdeauducq** [comment]
What messages does it emit when the connection fails?
How is the exception handled?

**Comment by Deepskyhunter** [comment]
> What messages does it emit when the connection fails?
How is the exception handled?

It will state the termination by the ctlmgr and print out the error in the log
Screenshot from 2022-06-24 12-03-36

**Comment by sbourdeauducq** [comment]
This does not answer the second question.

**Comment by Deepskyhunter** [comment]
And will restart by the ctlmgr

**Comment by sbourdeauducq** [comment]
Please revert [6561dd4](https://github.com/m-labs/artiq/pull/1924/commits/6561dd426a70e4c04f34138dfb2efbe489c9bf5c) and [120658d](https://github.com/m-labs/artiq/pull/1924/commits/120658d6f6fce98f02b8f343bf2ef5f957606b45), then answer the second question before writing any further code.

File: artiq/frontend/aqctl_corelog.py [hunk 1]
```diff
@@ -115,6 +113,9 @@ def main():
                     loop.run_until_complete(get_logs_task)
                 except asyncio.CancelledError:
                     pass
+        except Exception as e:
+            logger.error(e)
```
Comment (sbourdeauducq) [inline artiq/frontend/aqctl_corelog.py:117]: No. 
Notice that other code uses ``exc_info``, there's a reason for that.

File: artiq/frontend/aqctl_corelog.py [hunk 2]
```diff
@@ -51,6 +59,7 @@ async def get_logs(host):
     await writer.drain()
 
     while True:
+
```
Comment (sbourdeauducq) [inline artiq/frontend/aqctl_corelog.py:62]: ?

File: artiq/frontend/aqctl_corelog.py [hunk 3]
```diff
@@ -101,6 +113,8 @@ def main():
                     loop.run_until_complete(get_logs_task)
                 except asyncio.CancelledError:
                     pass
+        except:
+            logger.error("Lost connection to the core device due to: ", exc_info = True)
```
Comment (sbourdeauducq) [inline artiq/frontend/aqctl_corelog.py:117]: Just say "Terminating due to exception". There can be exceptions raised for reasons other than the core device connection being lost - in fact one can even be raised (and this code executed) before the core device has been contacted at all. So your message is misleading.
Remove the spaces around =

File: artiq/frontend/aqctl_corelog.py [hunk 4]
```diff
@@ -89,7 +98,10 @@ def main():
...
                 try:
                     _, pending = loop.run_until_complete(asyncio.wait(
-                        [signal_handler.wait_terminate(), server.wait_terminate()],
+                        [signal_handler.wait_terminate(),
+                         server.wait_terminate(),
+                         get_logs_task
+                        ],
                         return_when=asyncio.FIRST_COMPLETED))
                     for task in pending:
                         task.cancel()
```
Comment (sbourdeauducq) [inline artiq/frontend/aqctl_corelog.py:107]: You're already cancelling ``get_log_task`` here. Don't cancel it again below....

**Comment by sbourdeauducq** [comment]
Actually this looks problematic: if ``get_logs_task`` terminates with exception then nothing is logged. Have you tested it?

**Pull request merged**